Feature/add socials

AlouchLaBouche opened the pull request:
In this PR, we are now able to add socials while creating your account! There is a list of socials that the user is able to choose from defined in `User.kt`, from which he can add his desired socials, and the corresponding username/phone number/website.
These socials will be directly visible in his User Profile page.

This PR includes:
- The new `SocialOverlay.kt` UI.
- Error handling in `AccountDetails.kt` and `SocialOverlay.kt` the user is now required to fill in a first and last name before creating an account.
- 2 new social logos, linked to the UI.
- A fix concerning the handling of empty uids in `FirestoreReferenceList.kt`.
- 2 new methods in the `UserViewModel`, abeling us to add new users.
- The corresponding UI and Unit tests.

What needs to be worked  on: 
- The error handling in `SocialOverlay.kt` is very basic, for instance, the only thing checked when adding a URL for the website is if it starts with `https://`. This check is far from ideal, as it restricts a lot of websites (especially personal ones) but at least ensures a certain level of security. An alternative proposed by @oskar-codes could be to add a warning pop up just like in discord to warn the user of potential harm when leaving the application. 
- Quality of life features, such as triming the username, being able to input different types of phone numbers in different formats ect...

oskar-codes commented on app/src/main/java/com/android/unio/ui/authentication/AccountDetails.kt:
```diff
@@ -159,33 +180,60 @@ fun AccountDetails(
-                      hasProvidedAccountDetails = true)
-              uploadUser(user, userRepositoryFirestore, navigationAction, context)
-              navigationAction.navigateTo(Screen.HOME)
+              if (firstName.isEmpty() ||
```
Consider extracting validation logic in separate functions.

Aurelien9Code replied:
Minor fix : Consider reloading error messages when clicking again on the create button, as I had an issue when running this part :  
I willingly did not put a name and a surname so they became red and ask me to fill them both before continuing. I did fullfilled the name and then click again on the create button. However, both boxes were still in red.

It can lead to confusion for the user. When they correct their input and attempt to proceed, they expect the UI to reflect their changes. If the fields remain highlighted in red, it suggests that the input is still invalid, causing frustration and potentially leading users to believe there is a bug or that their input is still incorrect. Implementing a refresh of the error states upon each attempt to submit would enhance user experience by providing immediate feedback and clarity on the input status.

oskar-codes commented on app/src/main/java/com/android/unio/ui/authentication/AccountDetails.kt:
```diff
@@ -159,33 +180,60 @@ fun AccountDetails(
+                  lastName.isEmpty() ||
+                  lastName.isBlank()) {
+                isError = true
+                errorMessage = "Please fill in your name"
```
Consider creating an error enum with messages, instead of writing it here.

oskar-codes reviewed (approved):
This effectively adds the last few touches to the sign-in process, great work on this!

AlouchLaBouche commented:
Thank your for reviewing my code @oskar-codes. I'll quickly go fix the problems you spotted 👍.

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/User.kt:
As explained in the PR string management, you should put the UserSocialError in Strings.kt as it is only a string handling with constants

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/User.kt:
```diff
@@ -58,3 +58,55 @@ data class User(
+
+  when (userSocial.social) {
+    Social.WHATSAPP -> {
+      val numberRegex = Regex("^[0-9]{10,12}")
```
The regex for phone numbers might be too strict or not account for all valid formats across different countries. Consider allowing optional country codes or using a more flexible regex pattern.

Aurelien9Code replied:
update : as @oskar-codes told me, it's actually a requirement from WhatsApp. Sorry, the placeholder you made is actually very clear.

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/User.kt:
```diff
@@ -17,15 +17,15 @@ enum class Interest(val title: String) {
-  WEBSITE("Website", R.drawable.website_icon),
-  OTHER("Other", R.drawable.other_icon)
+enum class Social(val title: String, val icon: Int, val URL: String, val URLshort: String) {
+  FACEBOOK("Facebook", R.drawable.facebook_icon, "https://www.facebook.com/", "facebook.com/"),
```
Is Facebook really useful for a student-aged application in 2024 ?

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/UserViewModel.kt:
```diff
@@ -32,6 +32,10 @@ class UserViewModel(val repository: UserRepository, initializeWithAuthenticatedU
   fun getUserByUid(uid: String, fetchReferences: Boolean = false) {
+    if (uid.isEmpty()) {
+      return
+    }
```
Consider logging a warning when an empty uid is encountered. This can help with debugging and understanding user flow.

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/UserViewModel.kt:
```diff
@@ -66,6 +70,21 @@ class UserViewModel(val repository: UserRepository, initializeWithAuthenticatedU
+        onSuccess = onSuccess,
+        onFailure = { Log.e("UserViewModel", "Failed to add user", it) })
+    _user.value = user
+  }
```
Minor fix : We could handle the case where the update might fail due to network issues or validation errors, and provide appropriate user feedback through a Toast.

Aurelien9Code commented on app/src/main/java/com/android/unio/model/user/UserViewModel.kt:
Great work ! It makes the UserViewModel class more functional and robust

Aurelien9Code commented on app/src/main/java/com/android/unio/ui/authentication/AccountDetails.kt:
Maybe as @oskar-codes  said, consider breaking down the AccountDetails composable into smaller functions or components, as it would be more readable than a big bloc.

Aurelien9Code replied:
Minor fix : make sure to handle any potential data loss or input validation when the user navigates back and forth between overlays

Aurelien9Code reviewed (approved):
Overall very good code and very good new functionalities for the project ! 🚀🚀🚀
And I personaly think that adding socials is the perfect way for people to be even more linked to other, in order to become by ourselves a social media.

However there is a several numbers of minor fixes you should take in consideration. I will put the ones you did not do in small tasks in the Scrum Board : )

AlouchLaBouche commented:
Thanks for your review @oskar-codes and @Aurelien9Code. I addressed most of the issues in my latest commit noticeably : 
- Incorrect Regex pattern.
- Better error handling when the user inputs his first and last name.
- Removed unnecessary comments.
- A companion object that appends `https://` for socials.

You are both correct about the poor handling of phone number and websites. But I believe that these changes can be done in another task as my PR is big enough as it is 😓.
